File: pset1/credit.c
#include <stdio.h>
#include <cs50.h>

//made by edye lautter

long get_number(void);
int count_number(long n);
bool luh_alg(long n, int size);
string get_flag(long n);

int main(void)
{
    long n = get_number();
    int size = count_number(n);

    if (size != 13 && size != 15 && size != 16)
    {
        printf("INVALID\n");
    }
    else
    {
        if (luh_alg(n, size))
        {
            printf("%s", get_flag(n));
        }

        else
        {
            printf("INVALID\n");
        }
    }

    return 0;
}

string get_flag(long n)
{
    while (n > 100)
    {
        n /= 10;
    }

    if (n / 10 == 4)
    {
        return "VISA\n";
    }
    else if (n == 34 || n == 37)
    {
        return "AMEX\n";
    }
    else if (n == 51 || n == 52 || n == 53 || n == 54 || n == 55)
    {
        return "MASTERCARD\n";
    }
    else
    {
        return "INVALID\n";
    }
}

bool luh_alg(long n, int size)
{
    long mult = 1;
    long sum = 0;
    long sum1 = 0;
    long temp;
    for (int i = 0; i < size; i++)
    {
        temp = 0;
        if (i % 2 != 0)
        {
            temp = 2 * ((n / mult) % 10);
            if (temp >= 10)
            {
                temp = temp % 10 + temp / 10;
            }

            sum += temp;

        }
        else
        {
            temp = (n / mult) % 10;
            sum1 += temp;
        }
        mult *= 10;
    }

    if ((sum + sum1) % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

int count_number(long n)
{
    int count = 0;
    while (n != 0)
    {
        n /= 10;
        count++;
    }
    return count;
}

long get_number(void)
{
    long n = get_long("Number: ");

    return n;
}

File: pset6/credit/credit.py
# trying to figure out the credit cards flags only by theyr numbers

from cs50 import get_string

def main():
    n = get_number()
    if len(n) != 13 and len(n) != 15\
        and len(n) != 16:
            print("INVALID")
    else:
        if luh_alg(n):
            get_flag(n)
        else:
            print("INVALID")

def luh_alg(num):
    mult = 1
    soma = 0
    soma1 = 0
    num_cpy = int(num)
    for i in range(len(num)):
        temp = int(num_cpy / mult) % 10
        if i % 2 != 0:
            temp = 2 * temp
            if temp >= 10:
                temp = (temp % 10) + int(temp / 10)
            soma += temp
        else:
            soma1 += temp
        mult *= 10
    #print(int(soma + soma1))
    if (int(soma + soma1) % 10) == 0:
        return True
    else:
        return False

def get_flag(number):
    l2num = number[:2]
    if l2num[:1] == "4":
        print("VISA")
    elif l2num == "34" or l2num == "37":
        print("AMEX")
    elif int(l2num) in range(51, 56):
        print("MASTERCARD")
    else:
        print("INVALID")
    #print(l2num)

def get_number():
    n = get_string("Number: ")
    return n

if __name__ == "__main__":
    main()
